### PR TITLE
vtk: Fix proj dependency ranges 

### PR DESCRIPTION
**Pull request by Chrismarsh**
`proj` had a breaking API change in version 8, when the deprecated `proj_api.h` was finally removed. The current range selection of proj in vtk would allow `vtk@9.0` to link against across this API change, e.g., to `proj@9`, and fail to compile.

More details on proj api backwards compatibility
https://proj.org/en/9.3/development/migration.html

This patch adds a new version constraint on the `proj` dependency to protect against this

@chuckatkins  @danlipsa
/cc @johnwparent

### CODE REVIEW

File: var/spack/repos/builtin/packages/vtk/package.py [hunk 1]
```diff
@@ -173,7 +173,8 @@ class Vtk(CMakePackage):
     depends_on("gl2ps", when="@8.1:")
     depends_on("gl2ps@1.4.1:", when="@9:")
     depends_on("proj@4", when="@8.2.0")
```
Comment (johnwparent) [inline var/spack/repos/builtin/packages/vtk/package.py]: I know it's not really part of this PR, but can we add the constraint on the lower bound as well? I.e. right now vtk@7 could match to proj 9

**Comment by johnwparent** [comment]
~Upstream VTK builds with (and vendors) Proj 8.1, so it should be possible to use VTK ( with newer projs. Lets try to implement whatever patching VTK is doing in the upstream here as not to extend the already fairly complex constraints between VTK and Proj (and other packages in that DAG).~
This PR is correct w.r.t. supported version of Proj corresponding to versions of VTK

File: var/spack/repos/builtin/packages/vtk/package.py [hunk 2]
```diff
@@ -173,7 +173,8 @@ class Vtk(CMakePackage):
...
-    depends_on("proj@4:", when="@9:")
+    depends_on("proj@4:7", when="@9.0")
+    depends_on("proj@8:", when="@9.1:")
```
Comment (johnwparent) [inline var/spack/repos/builtin/packages/vtk/package.py]: I think this needs to be 9.2, 9.1 still vendors proj@4.4

File: var/spack/repos/builtin/packages/vtk/package.py (hunk 1 above)
Comment (Chrismarsh) [reply var/spack/repos/builtin/packages/vtk/package.py]: fixed

File: var/spack/repos/builtin/packages/vtk/package.py (hunk 2 above)
Comment (Chrismarsh) [reply var/spack/repos/builtin/packages/vtk/package.py]: Fixed

**Review by johnwparent — DISMISSED** [review]

**Comment by Chrismarsh** [comment]
This is blocked until https://github.com/spack/spack/pull/44535 can be merged in.

**Comment by Chrismarsh** [comment]
@johnwparent I rebased this on `develop` as the CI failure should be fixed now https://github.com/spack/spack/pull/42767 is merged in. I couldn't get a clean push, so had to force push; sorry about that.

File: var/spack/repos/builtin/packages/vtk/package.py [hunk 3]
```diff
@@ -171,8 +171,10 @@ class Vtk(CMakePackage):
     depends_on("utf8cpp", when="@9:")
     depends_on("gl2ps", when="@8.1:")
     depends_on("gl2ps@1.4.1:", when="@9:")
-    depends_on("proj@4", when="@8.2.0")
-    depends_on("proj@4:", when="@9:")
+    depends_on("proj@4", when="@:8.2.0")
+    depends_on("proj@4:7", when="@:8.2.0")
```
Comment (johnwparent) [inline var/spack/repos/builtin/packages/vtk/package.py]: Sorry, should have caught this earlier, but doesn't this now imply the `depends_on` above it?

**Review by johnwparent — APPROVED** [review]

File: var/spack/repos/builtin/packages/vtk/package.py [hunk 4]
```diff
@@ -171,8 +171,9 @@ class Vtk(CMakePackage):
     depends_on("utf8cpp", when="@9:")
     depends_on("gl2ps", when="@8.1:")
     depends_on("gl2ps@1.4.1:", when="@9:")
-    depends_on("proj@4", when="@8.2.0")
-    depends_on("proj@4:", when="@9:")
+    # "8.2.1a" uses an internal proj so this special cases 8.2.1a
+    depends_on("proj@4:7", when="@:8.2.0, 9:9.1")
```
Comment (johnwparent) [inline var/spack/repos/builtin/packages/vtk/package.py:175]: TIL this syntax is a thing

**Comment by jcortial-safran** [comment]
I never had any problem building VTK@9.1 with proj@9.2.1. I am not sure it was necessary to forbid that particular combination.

**Comment by johnwparent** [comment]
@jcortial-safran 
This PR has been merged for ~1 year, in the future please file an issue or post in the slack if you're experiencing issues or would like behavior changed and link to the merged PR. 

> I never had any problem building VTK@9.1 with proj@9.2.1. I am not sure it was necessary to forbid that particular combination.

Vtk 9.1 vendors proj 4.4 but does look like it has some mitigations for the breaking change in proj 8. I'll PR a relaxation of the version constraints and we'll see how CI does.

**Pull request merged**